### DISCUSSION TITLE
Does the pico support virtual memory, for example with an external spi SD card

### DISCUSSION THREAD

**Discussion by FlydGitHub** [micropython/micropython#10921] [RP2040 / Pico]
Does the pico support virtual memory, for example with an external spi SD card?

**Comment by niftyhacking** [comment]
On Fri, Mar 3, 2023 at 15:06 FlydGitHub ***@***.***> wrote:

> Does the pico support virtual memory, for example with an external spi SD
> card?
>
Three things are mixed here:
 - vm generally the operating system (not on a microcontroller)
 - garbage collection (see link)
 - external storage.
Since there is no operating system supporting micro python the simple
answer is no.
It is reasonable to build access to external storage and move data in and
out under application control. There would not be a filesystem, or on
demand page fault handler.

https://docs.micropython.org/en/latest/library/gc.html

A lot is possible. “ RP2040 is a stateless device, with support for cached
execute-in-place from external QSPI memory.” On the pico.

Dig into the processor starting here.

https://www.raspberrypi.com/documentation/microcontrollers/rp2040.html#welcome-to-rp2040
And here.
https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf

So no but interesting things are very possible.

This is informative:
https://www.adafruit.com/product/4864

Raspberry Pi Pico and RP2040 FAQ

   - Should I buy a Raspberry Pi Pico or a Raspberry Pi Zero?
   These are two very different devices, a microcontroller and a
   microcomputer, intended for different purposes. Pico is a great device for
   dedicating to a very specific task, whereas Zero is a multipurpose device.
   Raspberry Pi Zero has HDMI out, a camera interface, etc; Raspberry Pi Pico
   does not. However, Pico has an on-board ADC as well as other peripherals
   not present on Raspberry Pi Zero, and consumes considerably less power; it
   is therefore much more suited to embedded applications than Raspberry Pi
   Zero.


-- 
Tinny keyboard.. Mobile ... I am

**Comment by peterhinch** [comment]
Virtual memory - the use of a filesystem to page blocks of RAM out to disk - is a concept alien to microcontrollers. It requires a hardware memory management unit (MMU) which microcontrollers don't possess. Some microcontrollers support external SPIRAM chips which can expand system RAM. Commonly they support XIP (execute in place) Flash memory which allows code to be stored in Flash and executed from there. MicroPython supports this with the use of frozen bytecode.
